libs/ml/include/ml/optimisation/optimiser.hpp

```diff
@@ -229,8 +231,36 @@ typename T::Type Optimiser<T, C>::Run(
   DataType                                     loss_sum{0};
   SizeType                                     step{0};
   std::pair<ArrayType, std::vector<ArrayType>> input;
+
+  high_resolution_clock::time_point cur_time;
+  high_resolution_clock::time_point prev_time = high_resolution_clock::now();
+  SizeType                          prevStep{0};
+
   while (!loader.IsDone() && step < subset_size)
   {
+    // print the training stats every batch
```

Comment:
you should set a flag in optimiser that defaults to not printing these statistics